<?xml version="1.0"?>
<launch>
    <arg name="sensor_model" description="VLP16|VLP32|VLS128"/>
    <arg name="return_mode" default="Dual" description="See readme for supported return modes"/>
    <arg name="frame_id"  default="velodyne"/>
    <arg name="scan_phase" default="0.0" />

    <arg name="calibration_file" default="$(find-pkg-share nebula_decoders)/calibration/velodyne/$(var sensor_model).yaml"/>

    <arg name="sensor_ip" default="192.168.1.201" description="Lidar Sensor IP"/>
    <arg name="host_ip" default="255.255.255.255" description="Broadcast IP from Sensor"/>
    <arg name="data_port" default="2368" description="LiDAR Data Port"/>
    <arg name="gnss_port" default="2369" description="LiDAR GNSS Port"/>
    <arg name="packet_mtu_size" default="1500" description="Packet MTU size"/>
    <arg name="rotation_speed" default="600" description="Motor RPM, the sensor's internal spin rate."/>
    <arg name="cloud_min_angle" default="0" description="Field of View, start degrees."/>
    <arg name="cloud_max_angle" default="360" description="Field of View, end degrees."/>
    <arg name="diag_span" default="1000" description="milliseconds"/>

    <arg name="setup_sensor" default="True" description="Enable sensor setup on hw-driver."/>

    <arg name="invalid_point_remove" default="true"/>

<!--    <arg name="invalid_rings" default="[0]"/>-->
<!--    <arg name="invalid_angles_start" default="[0]"/>-->
<!--    <arg name="invalid_angles_end" default="[0]"/>-->

    <arg name="invalid_rings" default="[12,0,1,2,3,4,5,6,0,1,2,3,4]"/>
    <arg name="invalid_angles_start" default="[31110,3500,3400,3200,3200,3000,2950,3100,26000,26000,26000,26000,26000]"/>
    <arg name="invalid_angles_end" default="[31495,6900,6500,4600,4600,4300,4050,3800,32200,32200,31000,30000,28000]"/>

    <node pkg="nebula_ros" exec="velodyne_driver_ros_wrapper_node"
          name="velodyne_cloud" output="screen">
        <param name="sensor_model" value="$(var sensor_model)"/>
        <param name="return_mode" value="$(var return_mode)"/>
        <param name="frame_id" value="$(var frame_id)"/>
        <param name="scan_phase" value="$(var scan_phase)"/>
        <param name="calibration_file" value="$(var calibration_file)"/>
        <param name="invalid_point_remove" value="$(var invalid_point_remove)"/>
        <param name="invalid_rings" value="$(var invalid_rings)"/>
        <param name="invalid_angles_start" value="$(var invalid_angles_start)"/>
        <param name="invalid_angles_end" value="$(var invalid_angles_end)"/>
    </node>

    <node pkg="nebula_ros" exec="velodyne_hw_ros_wrapper_node"
          name="velodyne_hw_driver" output="screen">
        <param name="sensor_model" value="$(var sensor_model)"/>
        <param name="return_mode" value="$(var return_mode)"/>
        <param name="frame_id" value="$(var frame_id)"/>
        <param name="scan_phase" value="$(var scan_phase)"/>
        <param name="sensor_ip" value="$(var sensor_ip)"/>
        <param name="frame_id" value="$(var frame_id)"/>
        <param name="host_ip" value="$(var host_ip)"/>
        <param name="data_port" value="$(var data_port)"/>
        <param name="gnss_port" value="$(var gnss_port)"/>
        <param name="packet_mtu_size" value="$(var packet_mtu_size)"/>
        <param name="rotation_speed" value="$(var rotation_speed)"/>
        <param name="cloud_min_angle" value="$(var cloud_min_angle)"/>
        <param name="cloud_max_angle" value="$(var cloud_max_angle)"/>
        <param name="setup_sensor" value="$(var setup_sensor)"/>
    </node>

    <node pkg="nebula_ros" exec="velodyne_hw_monitor_ros_wrapper_node"
          name="hesai_hw_monitor_$(var frame_id)" output="screen">
        <param name="sensor_model" value="$(var sensor_model)"/>
        <param name="return_mode" value="$(var return_mode)"/>
        <param name="frame_id" value="$(var frame_id)"/>
        <param name="scan_phase" value="$(var scan_phase)"/>
        <param name="sensor_ip" value="$(var sensor_ip)"/>
        <param name="frame_id" value="$(var frame_id)"/>
        <param name="host_ip" value="$(var host_ip)"/>
        <param name="data_port" value="$(var data_port)"/>
        <param name="gnss_port" value="$(var gnss_port)"/>
        <param name="packet_mtu_size" value="$(var packet_mtu_size)"/>
        <param name="rotation_speed" value="$(var rotation_speed)"/>
        <param name="cloud_min_angle" value="$(var cloud_min_angle)"/>
        <param name="cloud_max_angle" value="$(var cloud_max_angle)"/>
        <param name="setup_sensor" value="$(var setup_sensor)"/>
    </node>

</launch>
